# Exercise 2 – Euclidean Distance

Calculate the distance between a new entry and the existing values using the Euclidean distance formula.

### Given dataset

| Brightness | Saturation | Class |
|---:|---:|---|
| 40 | 20 | Red |
| 50 | 50 | Blue |
| 60 | 90 | Blue |
| 10 | 25 | Red |
| 70 | 70 | Blue |
| 60 | 10 | Red |
| 25 | 80 | Blue |

For demonstration, the new entry is **(45, 40)**.

### Formula

$$d = \\sqrt{(x_2-x_1)^2 + (y_2-y_1)^2}$$

In [1]:
#ch.sc.u4cse24145
import pandas as pd
import numpy as np

# Given dataset
data = pd.DataFrame({
    'Brightness': [40, 50, 60, 10, 70, 60, 25],
    'Saturation': [20, 50, 90, 25, 70, 10, 80],
    'Class': ['Red', 'Blue', 'Blue', 'Red', 'Blue', 'Red', 'Blue']
})

print('Given Dataset:')
display(data)

Given Dataset:


,Brightness,Saturation,Class
0,40,20,Red
1,50,50,Blue
2,60,90,Blue
3,10,25,Red
4,70,70,Blue
5,60,10,Red
6,25,80,Blue


## New entry

We use **Brightness = 45** and **Saturation = 40** as the new entry.

In [2]:
new_brightness = 45
new_saturation = 40

data['Distance'] = np.sqrt(
    (data['Brightness'] - new_brightness) ** 2 +
    (data['Saturation'] - new_saturation) ** 2
)

print(f'New entry: ({new_brightness}, {new_saturation})\n')
print(data.to_string(index=False))

New entry: (45, 40)

 Brightness  Saturation Class  Distance
         40          20   Red 20.615528
         50          50  Blue 11.180340
         60          90  Blue 52.201533
         10          25   Red 38.078866
         70          70  Blue 39.051248
         60          10   Red 33.541020
         25          80  Blue 44.721360


## Manual calculation for the first point

For **(40, 20)**:

$$d = \\sqrt{(45-40)^2 + (40-20)^2}$$

$$d = \\sqrt{25 + 400} = \\sqrt{425} \\approx 20.62$$

In [3]:
# Display the distance calculation for every point
for _, row in data.iterrows():
    b = row['Brightness']
    s = row['Saturation']
    d = row['Distance']
    print(f'({b}, {s}) - {row["Class"]}: {d:.2f}')

(40, 20) - Red: 20.62
(50, 50) - Blue: 11.18
(60, 90) - Blue: 52.20
(10, 25) - Red: 38.08
(70, 70) - Blue: 39.05
(60, 10) - Red: 33.54
(25, 80) - Blue: 44.72


In [4]:
# Sort points from nearest to farthest
result = data.sort_values('Distance').reset_index(drop=True)
print('Distances from nearest to farthest:')
display(result)

Distances from nearest to farthest:


,Brightness,Saturation,Class,Distance
0,50,50,Blue,11.180340
1,40,20,Red,20.615528
2,60,10,Red,33.541020
3,10,25,Red,38.078866
4,70,70,Blue,39.051248
5,25,80,Blue,44.721360
6,60,90,Blue,52.201533


## Optional kNN classification

Using **k = 3**, the three nearest points are selected and majority voting determines the predicted class.

In [5]:
k = 3
nearest = result.head(k)
predicted_class = nearest['Class'].mode()[0]

print('Three nearest neighbors:')
display(nearest)
print(f'Predicted class for ({new_brightness}, {new_saturation}): {predicted_class}')

Three nearest neighbors:


,Brightness,Saturation,Class,Distance
0,50,50,Blue,11.180340
1,40,20,Red,20.615528
2,60,10,Red,33.541020


Predicted class for (45, 40): Red


## Conclusion

Euclidean distance was calculated between the new entry **(45, 40)** and every existing Brightness/Saturation point. The smaller the distance, the more similar the point is to the new entry.